--- a/sysdep/linux/netlink.c
+++ b/sysdep/linux/netlink.c
@@ -13,6 +13,11 @@
 #include <sys/uio.h>
 #include <errno.h>
 
+/* RTAX_QUICKACK since Linux 3.11 / RTAX_CC_ALGO since Linux 4.0 */
+#ifndef LINUX_VERSION_CODE
+#include <linux/version.h>
+#endif
+
 #undef LOCAL_DEBUG
 
 #include "nest/bird.h"
@@ -89,11 +94,13 @@ static struct ea_class ea_krt_metrics[]
     .format = krt_bitfield_format,
     .empty = krt_bitfield_empty,
   },
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
   [RTAX_CC_ALGO] = {
     .name = "krt_congctl",
     .legacy_name = "Kernel.congctl",
     .type = T_STRING,
   },
+#endif /* RTAX_CC_ALGO since Linux 4.0 */
 #define KRT_METRIC_INT(_rtax, _name)	[_rtax] = { .name = "krt_" _name, .legacy_name = "Kernel." _name, .type = T_INT }
   KRT_METRIC_INT(RTAX_MTU, "mtu"),
   KRT_METRIC_INT(RTAX_WINDOW, "window"),
@@ -107,7 +114,9 @@ static struct ea_class ea_krt_metrics[]
   KRT_METRIC_INT(RTAX_INITCWND, "initcwnd"),
   KRT_METRIC_INT(RTAX_RTO_MIN, "rto_min"),
   KRT_METRIC_INT(RTAX_INITRWND, "initrwnd"),
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0)
   KRT_METRIC_INT(RTAX_QUICKACK, "quickack"),
+#endif /* RTAX_QUICKACK since Linux 3.11 */
 #ifdef RTAX_FASTOPEN_NO_COOKIE
   KRT_METRIC_INT(RTAX_FASTOPEN_NO_COOKIE, "fastopen_no_cookie"),
 #else
@@ -358,6 +367,7 @@ nl_request_dump_neigh(int af, int bridge
     .ndm.ndm_family = af,
   };
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
   if (bridge_id)
   {
     req.rta.rta_type = NDA_MASTER;
@@ -365,6 +375,7 @@ nl_request_dump_neigh(int af, int bridge
     req.master_id = bridge_id;
     req.nh.nlmsg_len = NLMSG_ALIGN(req.nh.nlmsg_len) + req.rta.rta_len;
   }
+#endif /* NDA_MASTER since Linux 3.16 */
 
   send(nl_scan.fd, &req, req.nh.nlmsg_len, 0);
   nl_scan.last_hdr = NULL;
@@ -512,12 +523,15 @@ static struct nl_want_attrs ifinfo_attr_
   [IFLA_INFO_DATA]= { 1, 0, 0 },
 };
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,3,0)
 #define BIRD_IFLA_BR_MAX (IFLA_BR_VLAN_FILTERING+1)
 
 static struct nl_want_attrs ifla_br_attr_want[BIRD_IFLA_BR_MAX] = {
   [IFLA_BR_VLAN_FILTERING] = { 1, 1, sizeof(u8) },
 };
+#endif /* IFLA_BR_VLAN_FILTERING since Linux 4.3 */
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,12,0)
 #define BIRD_IFLA_VXLAN_MAX (IFLA_VXLAN_LOCAL6+1)
 
 static struct nl_want_attrs ifla_vxlan_attr_want[BIRD_IFLA_VXLAN_MAX] = {
@@ -526,8 +540,14 @@ static struct nl_want_attrs ifla_vxlan_a
   [IFLA_VXLAN_LOCAL]	= { 1, 1, sizeof(ip4_addr) },
   [IFLA_VXLAN_LOCAL6]	= { 1, 1, sizeof(ip6_addr) },
 };
+#endif
+/* IFLA_VXLAN_ID since Linux 3.7 */
+/* IFLA_VXLAN_LEARNING since Linux 3.7 */
+/* IFLA_VXLAN_LOCAL since Linux 3.7 */
+/* IFLA_VXLAN_LOCAL6 since Linux 3.12 */
 
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0)
 #define BIRD_IFBRIDGE_MAX (IFLA_BRIDGE_VLAN_TUNNEL_INFO+1)
 
 static struct nl_want_attrs ifbridge_attr_want[BIRD_IFBRIDGE_MAX] UNUSED = {
@@ -542,6 +562,7 @@ static struct nl_want_attrs ifbridge_vla
   [IFLA_BRIDGE_VLAN_TUNNEL_VID]		= { 1, 1, sizeof(u16) },
   [IFLA_BRIDGE_VLAN_TUNNEL_FLAGS]	= { 1, 1, sizeof(u16) },
 };
+#endif /* IFLA_BRIDGE_VLAN_TUNNEL_{FLAGS,ID,INFO} since Linux 4.11 */
 
 
 #define BIRD_IFA_MAX  (IFA_FLAGS+1)
@@ -637,6 +658,7 @@ static struct nl_want_attrs rtm_attr_wan
 #endif
 
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0)
 #define BIRD_NDA_MAX  (NDA_SRC_VNI+1)
 
 static struct nl_want_attrs ndm_attr_want[BIRD_NDA_MAX] = {
@@ -647,6 +669,7 @@ static struct nl_want_attrs ndm_attr_wan
   [NDA_MASTER]	  = { 1, 1, sizeof(u32) },
   [NDA_SRC_VNI]	  = { 1, 1, sizeof(u32) },
 };
+#endif /* NDA_SRC_VNI since Linux 4.11 */
 
 
 static int
@@ -712,10 +735,12 @@ static inline mac_addr rta_get_mac(struc
 static inline ip_addr rta_get_via(struct rtattr *a)
 {
   struct rtvia *v = RTA_DATA(a);
+#ifdef HAVE_MPLS_KERNEL
   switch(v->rtvia_family) {
     case AF_INET:  return ipa_from_ip4(ip4_ntoh(*(ip4_addr *) v->rtvia_addr));
     case AF_INET6: return ipa_from_ip6(ip6_ntoh(*(ip6_addr *) v->rtvia_addr));
   }
+#endif
   return IPA_NONE;
 }
 
@@ -1068,9 +1093,11 @@ nl_add_metrics(struct nlmsghdr *h, uint
 
   for (t = 1; t < max; t++)
     if (metrics[0] & (1 << t))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
       if (t == RTAX_CC_ALGO)
         nl_add_attr_str(h, bufsize, t, cc_algo);
       else
+#endif
         nl_add_attr_u32(h, bufsize, t, metrics[t]);
 
   nl_close_attr(h, a);
@@ -1091,6 +1118,7 @@ nl_parse_metrics(struct rtattr *hdr, u32
     if (a->rta_type >= max)
       continue;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
     if (a->rta_type == RTAX_CC_ALGO)
     {
       *cc_algo = rta_get_str(a);
@@ -1103,6 +1131,7 @@ nl_parse_metrics(struct rtattr *hdr, u32
       metrics[a->rta_type] = 0;
     }
     else
+#endif
     {
       if (RTA_PAYLOAD(a) != 4)
 	return -1;
@@ -1167,6 +1196,7 @@ nl_parse_link(struct nlmsghdr *h, int sc
     if (li[IFLA_INFO_KIND])
       kind = RTA_DATA(li[IFLA_INFO_KIND]);
 
+#ifdef CONFIG_BRIDGE
     if (!bstrcmp(kind, "bridge") && li[IFLA_INFO_DATA])
     {
       ea_set_attr_u32(&f.attrs, &ea_iface_type, 0, IF_TYPE_BRIDGE);
@@ -1208,6 +1238,7 @@ nl_parse_link(struct nlmsghdr *h, int sc
 	ea_set_attr_data(&f.attrs, &ea_iface_vxlan_ip_addr, 0, &addr, sizeof(addr));
       }
     }
+#endif
   }
 
   ifi = if_find_by_index(i->ifi_index);
@@ -1696,9 +1727,11 @@ nl_send_route(struct krt_proto *p, const
     int id = ea->id - ea_krt_metrics[0].id;
     metrics[0] |= 1 << id;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
     if (id == RTAX_CC_ALGO)
       cc_algo = ea->u.ptr->data;
     else
+#endif
       metrics[id] = ea->u.data;
   }
 
@@ -2096,9 +2129,11 @@ nl_parse_route(struct nl_parse_state *s,
 
       for (uint t = 1; t < KRT_METRICS_MAX; t++)
 	if (metrics[0] & (1 << t))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
 	  if (t == RTAX_CC_ALGO)
 	    ea_set_attr(&ra, EA_LITERAL_STORE_ADATA(&ea_krt_metrics[t], 0, cc_algo, strlen(cc_algo) + 1));
 	  else
+#endif
 	    ea_set_attr(&ra, EA_LITERAL_EMBEDDED(&ea_krt_metrics[t], 0, metrics[t]));
     }
 
