--- a/sysdep/linux/netlink.c
+++ b/sysdep/linux/netlink.c
@@ -13,6 +13,11 @@
 #include <sys/uio.h>
 #include <errno.h>
 
+/* RTAX_QUICKACK since Linux 3.11 / RTAX_CC_ALGO since Linux 4.0 */
+#ifndef LINUX_VERSION_CODE
+#include <linux/version.h>
+#endif
+
 #undef LOCAL_DEBUG
 
 #include "nest/bird.h"
@@ -89,11 +94,13 @@ static struct ea_class ea_krt_metrics[]
     .format = krt_bitfield_format,
     .empty = krt_bitfield_empty,
   },
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
   [RTAX_CC_ALGO] = {
     .name = "krt_congctl",
     .legacy_name = "Kernel.congctl",
     .type = T_STRING,
   },
+#endif /* RTAX_CC_ALGO since Linux 4.0 */
 #define KRT_METRIC_INT(_rtax, _name)	[_rtax] = { .name = "krt_" _name, .legacy_name = "Kernel." _name, .type = T_INT }
   KRT_METRIC_INT(RTAX_MTU, "mtu"),
   KRT_METRIC_INT(RTAX_WINDOW, "window"),
@@ -107,7 +114,9 @@ static struct ea_class ea_krt_metrics[]
   KRT_METRIC_INT(RTAX_INITCWND, "initcwnd"),
   KRT_METRIC_INT(RTAX_RTO_MIN, "rto_min"),
   KRT_METRIC_INT(RTAX_INITRWND, "initrwnd"),
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0)
   KRT_METRIC_INT(RTAX_QUICKACK, "quickack"),
+#endif /* RTAX_QUICKACK since Linux 3.11 */
 #ifdef RTAX_FASTOPEN_NO_COOKIE
   KRT_METRIC_INT(RTAX_FASTOPEN_NO_COOKIE, "fastopen_no_cookie"),
 #else
@@ -528,6 +537,7 @@ static struct nl_want_attrs ifla_vxlan_a
 };
 
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0)
 #define BIRD_IFBRIDGE_MAX (IFLA_BRIDGE_VLAN_TUNNEL_INFO+1)
 
 static struct nl_want_attrs ifbridge_attr_want[BIRD_IFBRIDGE_MAX] UNUSED = {
@@ -542,6 +552,7 @@ static struct nl_want_attrs ifbridge_vla
   [IFLA_BRIDGE_VLAN_TUNNEL_VID]		= { 1, 1, sizeof(u16) },
   [IFLA_BRIDGE_VLAN_TUNNEL_FLAGS]	= { 1, 1, sizeof(u16) },
 };
+#endif /* IFLA_BRIDGE_VLAN_TUNNEL_{FLAGS,ID,INFO} since Linux 4.11 */
 
 
 #define BIRD_IFA_MAX  (IFA_FLAGS+1)
@@ -637,6 +648,7 @@ static struct nl_want_attrs rtm_attr_wan
 #endif
 
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0)
 #define BIRD_NDA_MAX  (NDA_SRC_VNI+1)
 
 static struct nl_want_attrs ndm_attr_want[BIRD_NDA_MAX] = {
@@ -647,6 +659,7 @@ static struct nl_want_attrs ndm_attr_wan
   [NDA_MASTER]	  = { 1, 1, sizeof(u32) },
   [NDA_SRC_VNI]	  = { 1, 1, sizeof(u32) },
 };
+#endif /* NDA_SRC_VNI since Linux 4.11 */
 
 
 static int
@@ -712,10 +725,12 @@ static inline mac_addr rta_get_mac(struc
 static inline ip_addr rta_get_via(struct rtattr *a)
 {
   struct rtvia *v = RTA_DATA(a);
+#ifdef HAVE_MPLS_KERNEL
   switch(v->rtvia_family) {
     case AF_INET:  return ipa_from_ip4(ip4_ntoh(*(ip4_addr *) v->rtvia_addr));
     case AF_INET6: return ipa_from_ip6(ip6_ntoh(*(ip6_addr *) v->rtvia_addr));
   }
+#endif
   return IPA_NONE;
 }
 
@@ -1068,9 +1083,11 @@ nl_add_metrics(struct nlmsghdr *h, uint
 
   for (t = 1; t < max; t++)
     if (metrics[0] & (1 << t))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
       if (t == RTAX_CC_ALGO)
         nl_add_attr_str(h, bufsize, t, cc_algo);
       else
+#endif /* RTAX_CC_ALGO since Linux 4.0 */
         nl_add_attr_u32(h, bufsize, t, metrics[t]);
 
   nl_close_attr(h, a);
@@ -1091,6 +1108,7 @@ nl_parse_metrics(struct rtattr *hdr, u32
     if (a->rta_type >= max)
       continue;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
     if (a->rta_type == RTAX_CC_ALGO)
     {
       *cc_algo = rta_get_str(a);
@@ -1103,6 +1121,7 @@ nl_parse_metrics(struct rtattr *hdr, u32
       metrics[a->rta_type] = 0;
     }
     else
+#endif /* RTAX_CC_ALGO since Linux 4.0 */
     {
       if (RTA_PAYLOAD(a) != 4)
 	return -1;
@@ -1696,9 +1715,11 @@ nl_send_route(struct krt_proto *p, const
     int id = ea->id - ea_krt_metrics[0].id;
     metrics[0] |= 1 << id;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
     if (id == RTAX_CC_ALGO)
       cc_algo = ea->u.ptr->data;
     else
+#endif /* RTAX_CC_ALGO since Linux 4.0 */
       metrics[id] = ea->u.data;
   }
 
@@ -2096,9 +2117,11 @@ nl_parse_route(struct nl_parse_state *s,
 
       for (uint t = 1; t < KRT_METRICS_MAX; t++)
 	if (metrics[0] & (1 << t))
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,0,0)
 	  if (t == RTAX_CC_ALGO)
 	    ea_set_attr(&ra, EA_LITERAL_STORE_ADATA(&ea_krt_metrics[t], 0, cc_algo, strlen(cc_algo) + 1));
 	  else
+#endif /* RTAX_CC_ALGO since Linux 4.0 */
 	    ea_set_attr(&ra, EA_LITERAL_EMBEDDED(&ea_krt_metrics[t], 0, metrics[t]));
     }
 
@@ -2208,8 +2231,10 @@ nl_send_fdb(const net_addr *n0, const rt
       nl_add_attr_u32(&r->h, rsize, NDA_VNI, nh->label[0]);
 
     eattr *ea = ea_find(e->attrs, &ea_gen_mpls_label);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0)
     if (ea)
       nl_add_attr_u32(&r->h, rsize, NDA_SRC_VNI, ea->u.data);
+#endif /* NDA_SRC_VNI since Linux 4.11 */
 
     r->n.ndm_state |= NUD_PERMANENT;
   }
@@ -2352,6 +2377,7 @@ nl_parse_fdb(struct nl_parse_state *s, s
       nh->label[0] = rta_get_u32(a[NDA_VNI]);
     }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,11,0)
     if (a[NDA_SRC_VNI])
     {
       struct kbr_vlan *vlan = kbr_find_vlan_by_vni(p, oif->index, rta_get_u32(a[NDA_SRC_VNI]));
@@ -2363,6 +2389,7 @@ nl_parse_fdb(struct nl_parse_state *s, s
 
       ((net_addr_eth *) &n)->vid = vid = vlan->vid;
     }
+#endif /* NDA_SRC_VNI since Linux 4.11 */
 
     tunnel = true;
   }
